enh: more realistic benchmark

zerothi opened the pull request:
I think it is vital that benchmarks are realistic. Using fixed size arrays is not realistic in my opinion.

This PR does many things:

- benchmarks should not rely on convergence, rather a fixed number of iterations is more appropriate, otherwise timings are irrelevant
- The problem of inconsistent iterations between the C and Fortran versions is fixed. It was an indexing error when calculating the cell for the `rhoarr` array (C is 0-based, F is 1-based). So they now run the same number of iterations, regardless of `M`.   
- Moved timings about to make them more comparable.   
- Ensured that all examples now use allocate/malloc
- Calls in C and F code are much more similar

For running benchmarks it is vital that `M` is so large that the 2D array goes beyond the L3 cache to match realistic cases.

certik commented:
@zerothi thanks for this!

I have an idea: why don't you leave `poisson2d/optimized.f90` and `poisson2d/optimized.c` as they are, but submit new files both for Fortran and C with your version. I actually want to have both. I want to see how long it takes when only the computation is performed, but I also want to see when you benchmark everything, including memory allocation, as you did. The same with the changes you made --- I actually would like to have both.

Regarding naming, we can even just do `version1.f90`, `version2.f90`, etc.

What do you think?

zerothi commented:
Good idea, give me some time :)

zerothi committed:
mnt: re-arranged codes and added many new versions

Benchmarks should not rely on convergence, rather a fixed
number of iterations is more appropriate.

The problem of inconsistent iterations between the C and Fortran versions
was also fixed because of an indexing error when calculating the cell
coordinates (C is 0-based, F is 1-based). So they now run the same number
of iterations, regardless of M.
Note, this issue still persists in 2 and 3, but not in 4, I can't seem
to figure out why.

Moved timings about to make them more comparable.

All different benchmarks write out the same format output to more easily
compare them.

The later versions are more realistic since they use allocate/malloc
which is more appropriate.

Heavily updated the README.md file to explain what is going on.

zerothi commented:
Please have a look at the updated look. It is now more complete and easier to extend with new versions!

zerothi commented:
Note that benchmarking the allocations is not really fair I think. In this case I have left all allocation timings out since that should be a benchmark by it-self.  
The reason for not doing it is that some things may be allocated on the stack which one can't easily time without complicating matters quite a bit.
